core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -74,18 +74,23 @@ private[kafka] object LogValidator extends Logging {
     }
   }
 
-  private[kafka] def validateOneBatchRecords(records: MemoryRecords) {
+  private[kafka] def getFirstBatchAndMaybeValidateNoMoreBatches(records: MemoryRecords, sourceCodec: CompressionCodec): RecordBatch = {
     val batchIterator = records.batches.iterator
 
     if (!batchIterator.hasNext) {
-      throw new InvalidRecordException("Compressed outer record has no batches at all")
+      throw new InvalidRecordException("Record batch has no batches at all")
     }
 
-    batchIterator.next()
+    val batch = batchIterator.next()
 
-    if (batchIterator.hasNext) {
-      throw new InvalidRecordException("Compressed outer record has more than one batch")
+    // if the format is v2 and beyond, or if the messages are compressed, we should check there's only one batch.
+    if (batch.magic() >= RecordBatch.MAGIC_VALUE_V2 || sourceCodec != NoCompressionCodec) {
```

Comment:
Could be simplified as 
```
if (batchIterator.hasNext && 
!(batch.magic < RecordBatch.MAGIC_VALUE_V2 && sourceCodec == NoCompressionCodec)) {
  throw new InvalidRecordException("Compressed outer record has more than one batch")
} else {
  batch
}
```

Reply:
I feel it is easier to read in this way, though a bit verbose.